### PR TITLE
fix: dont early exit loop when filtering duplicates from relationship updates

### PR DESCRIPTION
**Pull request by runspired**
reported by @wagenet 

Affects 4.0, 4.1, 4.2 so should be backported as a critical fix to all three.

TL;DR our previous fix for filtering duplicates in relationships was incorrect: #7786. It had the effect that the first time a duplicate is encountered the loop would exit and all entries in the new state array after that would be empty.

### deprecate support for filtering duplicate entries from relationships

We should deprecate (later assert) many relationships having duplicates, as we discard them anyway.

### we could RFC allowing duplicates with some constraints

Doing this work reminds me of the common ask to explicitly *allow* duplicates in collection relationships. I do believe with the new relationship infrastructure that this is something we could potentially allow. If anyone wants to take on an RFC the constraints are roughly:

- relationships that can have duplicate must specify both `inverse: null` and `allowDuplicates` as soon as soon as flag in their config.

```ts
@hasMany("task", { inverse: null, allowDuplicates: true }) tasks;
```

- relationships configured in this way would utilize the `index` alongside the identifier to handle add/remove.
- relationships configured in this way would be considered to have a meaningful order (typically we do not assume this for relationships because inverses can obstruct that information, though we do attempt it.)
- addition and removals stored in the diff would need to store an index, those indexes would likely need recalculated when a change is committed. **This may prove prohibitively difficult and should be explored prior to RFC.**
- we would need to consider whether `HasMany` would still be the decorator, as we likely would need to do a little extra work to ensure duplicates are supported by the presentation class and also to update the order internally if for instance a "sort" was applied.

### CODE REVIEW

File: packages/-ember-data/node-tests/docs/test-coverage.js [hunk 1]
```diff
@@ -24,7 +24,7 @@ QUnit.module('Docs coverage', function (hooks) {
   if (!process.env.REUSE_DOCS) {
     buildDocs();
   }
-  const docs = require('../../dist/docs/data.json'); // eslint-disable-line node/no-missing-require
```
[the remark this answers is not in the document]
Comment (runspired) [reply packages/-ember-data/node-tests/docs/test-coverage.js:27]: yep! cc @bmish for fix

File: packages/record-data/addon/-private/graph/index.ts [hunk 2]
```diff
@@ -266,6 +266,7 @@ export class Graph {
         assert(`Can only perform the operation updateRelationship on remote state`, isRemote);
         if (DEBUG) {
           // in debug, assert payload validity eagerly
+          // TODO add deprecations/assertion here for duplicates
```
[the remark this answers is not in the document]
Comment (runspired) [reply packages/record-data/addon/-private/graph/index.ts:269]: We don't but I wish we did. That said I do make a habit of searching for them and trying to pay them down :)

File: packages/-ember-data/node-tests/docs/test-coverage.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (bmish) [reply packages/-ember-data/node-tests/docs/test-coverage.js:27]: ?

**Review by snewcomer — APPROVED** [review]
Wonderful!

File: packages/-ember-data/node-tests/docs/test-coverage.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (runspired) [reply packages/-ember-data/node-tests/docs/test-coverage.js:27]: @bmish It looks like we were too aggressive in removing the unused require lint rule. When I ran auto fix here it resulted in an unexpected removal.

[the remark this answers is not in the document]
Comment (bmish) [reply packages/-ember-data/node-tests/docs/test-coverage.js:27]: ESLint in [removing](https://eslint.org/docs/user-guide/command-line-interface#--report-unused-disable-directives) the unused disable directives.

My guess is that some people have the `dist` folder locally and some don't causing the disable directive to be unused sometimes but used other times?

[the remark this answers is not in the document]
Comment (runspired) [reply packages/-ember-data/node-tests/docs/test-coverage.js:27]: @bmish right, cause that file is generated at build time only when we generate the docs. I think we need to deactivate that for this line and/or deactivate it generally if it can't be deactivated for one line only.

**Comment by runspired** [comment]
@nbibler I think you are potentially right
